File: C++/particle.cpp
#include "particle.hpp"

namespace particle {
    Distance::Distance(Particle* part1, Particle part2) {
        vec::Vec2 l1 = part1->location;
        vec::Vec2 l2 = part2.location;
        dy = l2.y - l1.y;
        dx = l2.x - l1.x;
        dsq = dx * dx + dy * dy;
        dr = std::sqrt(dsq);
        if (dr == 0) {
            dr = 0.01;
        }
    }

    Distance::Distance(Particle part1, Particle part2) {
        vec::Vec2 l1 = part1.location;
        vec::Vec2 l2 = part2.location;
        dy = l2.y - l1.y;
        dx = l2.x - l1.x;
        dsq = dx * dx + dy * dy;
        dr = std::sqrt(dsq);
        if (dr == 0) {
            dr = 0.01;
        }
    }

    vec::Vec2 attraction(Particle* part1, Particle part2) {
        Distance dist(part1, part2);
        float f;

        if (dist.dr > 1) {
            f = (GRAVRATE * part1->mass * part2.mass) / dist.dsq;
            //std::cout << part1->mass << std::endl;
            return vec::Vec2(
                (f * dist.dx / dist.dr) / part1->mass,
                (f * dist.dy / dist.dr) / part1->mass
            );
        } else {
            f = 0;
            //std::cout << part1->mass << std::endl;

            return vec::Vec2(
                (f * dist.dx / dist.dr) / part1->mass,
                (f * dist.dy / dist.dr) / part1->mass
            );
        }
    }

    unsigned long _RGB(int r,int g, int b) {
		return b + (g<<8) + (r<<16);
	}
	
	void setColor(Display *dis, GC g, std::vector<int> rgb) {
		XSetForeground(dis, g, _RGB(rgb[0], rgb[1], rgb[2]));
	}

    Particle::Particle(vec::Vec2 velocity, vec::Vec2 location, float mass, std::vector<int> color){
        this->velocity = velocity;
        this->location = location;
        this->mass     = mass;
        this->color    = color;
    }

    void Particle::draw(Display *dis, Window win, GC gc) {
        setColor(dis, gc, color);
        std::vector<int> nl = location.normalize(width, height);
        XDrawArc(dis, win, gc, nl[0], nl[1], mass * 10, mass * 10, 0, 23040);
        XFillArc(dis, win, gc, nl[0], nl[1], mass * 10, mass * 10, 0, 23040);
        //XDrawPoint(dis, win, gc, nl[0], nl[1]);
    }

    void Particle::move() {
        location.addVec(velocity);
    }

    void Particle::merge(Particle part){
        if (mass < part.mass) {
            rem = true;
            float newvx = (part.velocity.x * part.mass + velocity.x * part.mass) / (part.mass + mass);
            float newvy = (part.velocity.y * part.mass + velocity.y * part.mass) / (part.mass + mass);
            part.velocity.addVec(vec::Vec2(newvx, newvy));
            part.mass += mass;
        } else {
            part.rem = true;
            float newvx = (velocity.x * mass + part.velocity.x * mass) / (mass + part.mass);
            float newvy = (velocity.y * mass + part.velocity.y * mass) / (mass + part.mass);
            velocity.addVec(vec::Vec2(newvx, newvy));
            mass += part.mass;
        }
    }

    void Particle::grav(std::vector<Particle> particles) {
        for (Particle &part : particles) {
            Particle* part1p = &part;
            Particle* part2p = this;
            if (part1p != part2p) {
                velocity.addVec(particle::attraction(this, part));
            }
        }
    }
}

File: Python/particle.py
from dataclass import Velocity, Location, Distance, Center
import pygame
import math

GRAVRATE = 1.3
CENTER = Center()

def attraction(par1, par2) -> Velocity:
    '''GMm/R**2'''
    dist = Distance(par1, par2)
    f = (GRAVRATE * par1.mass * par2.mass) / dist.dsq if dist.dr > 1 else 0
    return Velocity(
        (f * dist.dx / dist.dr) / par1.mass,
        (f * dist.dy / dist.dr) / par1.mass
    )

class Particle:
    def __init__(self, 
                velocity = Velocity(1, 5), 
                location = Location(0, 0), 
                mass = 1, 
                color = (255, 255, 255)
            ):
        self.velocity = velocity
        self.location = location
        self.mass = mass
        self.color = color
        self.rem = False
    
    def draw(self, screen):
        width  = pygame.display.Info().current_w
        height = pygame.display.Info().current_h
        #if abs(self.location.normalize(width, height)[0]) > width or abs(self.location.normalize(width, height)[1]) > height:
        #    self.rem = True
        pygame.draw.circle(screen, self.color, self.location.normalize(width, height), self.mass * 2)
    
    def move(self):
        self.location.addVec(self.velocity)
    
    def Merge(self, part):
        # https://www.thanassis.space/gravity.html
        if self.mass < part.mass:
            p1, p2 = part, self
        else:
            p1, p2 = self, part

        p2.rem = True
        newvx = (p1.velocity.x*p1.mass+p2.velocity.x*p2.mass)/(p1.mass+p2.mass)
        newvy = (p1.velocity.y*p1.mass+p2.velocity.y*p2.mass)/(p1.mass+p2.mass)
        p1.velocity.addVec(Velocity(newvx, newvy))
        p1.mass += p2.mass

    def grav(self, particles):
        width  = pygame.display.Info().current_w
        height = pygame.display.Info().current_h
        for part in particles:
            if part is not self:
                self.velocity.addVec(attraction(self, part))
        '''
        if abs(self.location.normalize(width, height)[0]) > width or abs(self.location.normalize(width, height)[1]) > height:
            a = attraction(self, CENTER)
            dist = Distance(self, CENTER)
            self.velocity.addVec(Velocity(a.x * dist.dx, a.y * dist.dy))
        else:
            self.velocity.addVec(attraction(self, CENTER))
        '''